README: Update the local building section

ajelinski opened the pull request:
This is an update for the README section about building packages locally. The commands provided are sufficient to build each and every package in a freshly created `ubuntu:bionic` Docker container.

The only differences between this section from `-compilers` and the one from `-eda`, `-misc` and `-prog` repositories will be:
1. the repository name in lines 130-132 (from the new file),
2. the "additional env vars" part https://github.com/litex-hub/litex-conda-compilers/pull/15/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R144, which in `-eda` can be much shorter:

    > Currently required additional environment variables are:
    > * `LIBFFI_VERSION` (`syn/symbiflow-yosys` only) – must contain a valid
    > version of `libffi` Conda package, eg. `3.3`.

    while in `-misc` and `-prog` I want to replace it with:

    > Currently there are no such variables and recipes in this repository,
    > although there are a few such cases in other `litex-conda-*` repositories.

Fixes litex-hub/litex-conda-eda#9
Fixes litex-hub/litex-conda-eda#10 (also thanks to the previous recipe changes in all four repositories)
Fixes litex-hub/litex-conda-eda#52
Fixes litex-hub/litex-conda-eda#53

mithro commented on README.md:
```diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+
+Almost all of the recipes require exporting `DATE_NUM` and `DATE_STR`
+variables, since those are used as a build number and a build string
+respectively for the created package.
```
```suggestion
respectively for the created package.

```

mithro commented on README.md:
```diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+  * `riscv64`,
+  * `sh`.
+
+Almost all of the recipes require exporting `DATE_NUM` and `DATE_STR`
```
`conda-build-prepare` should detect or produce reasonable default values for these so a user doesn't need to set them when building locally.

mithro commented on README.md:
```diff
@@ -123,27 +176,49 @@ get_built_package() {
+The tool is also used within the CI, which makes the locally built
+package much more similar to the one built by the CI workflow.
+
+The `WORKDIR` variable is the name of the directory that will be
```
I don't like the name `WORKDIR` -- I would prefer something more descriptive like `FINALIZE_PACKAGE_OUTPUTDIR` or something along those lines?

ajelinski replied:
I've changed it to `PREPARED_RECIPE_OUTPUTDIR` since it's a path of the new directory that will contain output after preparing the recipe. Of course I'm open for modifications if you think it should stress the fact that it's finalized or something else. I think no name is fully accurate apart from the longish and less descriptive `CONDA_BUILD_PREPARE_OUTPUTDIR. Perhaps `CBP_OUTPUTDIR` could be used but it's still less descriptive.

mithro commented on README.md:
```diff
@@ -123,27 +176,49 @@ get_built_package() {
+
+The `WORKDIR` variable is the name of the directory that will be
+created  as an output of running `conda-build-prepare`.
+It will contain these subdirectories:
```
Can you explain what is in each of these directories?

mithro commented on README.md:
````diff
@@ -123,27 +176,49 @@ get_built_package() {
-cd conda-hdmi2usb-packages
+```bash
+# Prepare the RECIPE with `conda-build-prepare`
+ADDITIONAL_PACKAGES="conda-build=3.20.3 conda-verify jinja2 pexpect python=3.7"
````
I'm confused about why `ADDITIONAL_PACKAGES` is needed?

ajelinski replied:
This is what we use in the CI so if we seek the maximum similarity with the CI then locally these should also be used.
I'm quite sure that only these set with versions make any difference so the rest (`conda-verify jinja2 pexpect`) can be dumped both from the CI and from here. Should I do it?

ajelinski replied:
In the end I've removed `conda-verify jinja2 pexpect` from here as I'm pretty sure they have zero influence on building.

`conda-build` and `python` have been left with versions for maximal CI resemblence.

mithro commented on README.md:
```diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
-Given a fairly empty *disposable* Ubuntu x86-64 test environment (eg,
-created with Docker, or Vagrant), it is possible to simulate *part* of
-what Travis CI will do to test building individual toolchain architectures
+Given a fairly empty *disposable* Ubuntu x86-64 test environment (e.g.,
```
It is recommend to build these packages in a fresh disposable environment such as a clean docker container.

While the goal is for the conda environment to be totally self contained, there is a constant battle to make sure this happens.

mithro commented on README.md:
```diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+* [Git](https://git-scm.com/),
+* Conda installed and initialized, e.g., using
+[Miniconda](https://docs.conda.io/en/latest/miniconda.html)
+(it automatically makes required *python3* with *pip* and *setuptools*
```
```suggestion
(which includes self contained versions of the required *python3* with *pip* and *setuptools* tooling)
```

mithro commented on README.md:
```diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+(it automatically makes required *python3* with *pip* and *setuptools*
+available),
+* [conda-build-prepare](https://github.com/litex-hub/conda-build-prepare),
+* this repository cloned to a local directory.
```
```suggestion
this repository cloned to a local directory.
```

ajelinski replied:
I'm not sure you really meant because simply placing it after the list didn't look great. I started it with `and of course this repository...` to make it more natural.

mithro commented on README.md:
```diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+eval "$("$CONDA_PATH/bin/conda" "shell.bash" 'hook' 2> /dev/null)"
+
+# Install `conda-build-prepare`
+python3 -m pip install git+https://github.com/litex-hub/conda-build-prepare@v0.1#egg=conda-build-prepare
```
I wonder if this should be `pip install -e .` to use the locally cloned conda-build-prepare?

ajelinski replied:
I think it would only make sense if the `conda-build-prepare` will be modified and that's not the case for the most of the people building packages locally.

I guess a person who wants to modify a Python module knows or can easily find how to install it in the "editable" mode.

umarcor replied:
I agree with @ajelinski. Using `-e` is misleading for less experienced users, and the target developers/collaborators are likely aware of it. It might be added as a comment, but IMHO it should not be the default.

mithro commented on README.md:
```diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
-}
+### Required environment variables
+
+Most importantly, the `RECIPE` variable must be set with the appropriate
```
This isn't quite true. The `RECIPE` argument is given to conda-build-prepare command as an argument?

ajelinski replied:
You're definitely right. I've redesigned this subsection to separate variables used in the script from information about the CBP arguments.

mithro commented on README.md:
````diff
@@ -89,31 +89,84 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+```bash
+# Set required variables with some values if they haven't been set already
+RECIPE=${RECIPE:-binutils}
+if [ "$RECIPE" = "binutils"         \
````
I don't like this if statement here.

mithro reviewed (commented):
Lots of minor suggestions.

Overall, looking like it is heading in a great direction!

ajelinski commented:
I've modified it quite significantly. You were right that the required variables part was confusing. I hope it's much more precise right now.

The `DATE_NUM` and `DATE_STR` will no longer be required after merging this PR: https://github.com/litex-hub/conda-build-prepare/pull/3 . I will make sure this is merged before merging this PR and I'll change the revision so the latest cbp is installed using the provided commands.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
-what Travis CI will do to test building individual toolchain architectures
-locally.
+It is recommend to build these packages in a fresh disposable environment
+such as a clean docker container.
```
I suggest: "*such as a clean container (Docker, Podman...)*".

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
 
-This can be done with something like:
+The commands from the following subsections are enough to build a package
+in a clean docker container based on the *ubuntu* Docker image.
```
Is "any" version of Ubuntu supported? Or is is bionic only, for now?

I suggest using "*container image*" instead of "*Docker image*".

ajelinski replied:
Changed to "The commands from the following subsections were tested to be enough to build a package in a clean container based on `ubuntu` (`16.04`-`20.04`) or `debian` (`8`-`10`) Docker image." after I tested them on these OSs.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+(which includes self contained versions of the required *python3* with
+*pip* and *setuptools* tooling),
+* [conda-build-prepare](https://github.com/litex-hub/conda-build-prepare)
+and of course this repository cloned to a local directory.
```
I would try avoiding "*of course*" and such. I also use them more than I should, but they don't help much.

I suggest:

"*Apart from cloning this repository to a local directory, the following prerequisites are required:*"

ajelinski replied:
You're right, it's much better.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+* [conda-build-prepare](https://github.com/litex-hub/conda-build-prepare)
+and of course this repository cloned to a local directory.
+
+All of these requirements can be satisfied using the following commands:
```
"*On Debian/Ubuntu, these requirements can be satisfied...*"

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+
+# Download Miniconda and install in CONDA_PATH
+CONDA_PATH=~/conda
+wget -c https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
```
I would write L120-L122 in a single line:

```sh
curl -fsSL https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh | sh -s -- -p $CONDA_PATH -b -f 
```

ajelinski replied:
Generally it's a nice idea but unfortunately this script prohibits sourcing it. It can only be executed so this doesn't work.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
 
+### Required environment variables
+
+The recipe might require exporting some additional environment variables.
```
Which recipe? "The Conda recipe"? Which one?

[These are socratic questions]

ajelinski replied:
I've changed it to "Some recipes require exporting additional environment variables."

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+with one of the commands from the next subsection.
+
+Currently required additional environment variables are:
+* `TOOLCHAIN_ARCH` (by: `binutils`, `gcc/*`, `gdb` and `toolchain/linux-musl`
```
Maybe not as part of this PR, but this feels like a good candidate for a `configure` shell script.

ajelinski replied:
`TOOLCHAIN_ARCH` isn't about the host architecture so I'm not sure what would be the purpose of such a script in this case.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+The optional `DATE_NUM` and `DATE_STR` environment variables are used by the
+recipes to set the `build/number` and `build/string` keys during preparation.
+In case they're unset, recipe preparation provides them with default values
+based on the latest commit's committer date from the repository holding the
```
Some users might attempt to download an archive of the repos, instead of cloning them through git. If having the git history is required (which I find sensible), I believe it'd be interesting to specify it above.

ajelinski replied:
You're right. I didn't think enough about downloading the repository as an archive. I stressed such a necessity in the description and added the commented `export DATE_*` commands to the script.

I will consider providing them automatically even if there's no git repository, as this can be more common than I thought initially.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+
+### Preparing and building the package
+
+After getting all prerequisites and setting the required variables, it
```
I feel there might be misunderstandings about what "local" means. There are three possible setups:

- The "native" Ubuntu environments.
- A container.
- Executing some commands natively and others in a container.

Those three setups are possible either locally or in CI. Hence, 6 total. At first, I thought that you were only considering "native Ubuntu environments" in CI and "a container" locally. Then, I understood that repos might be cloned natively, and other commands executed in a container. Now, I am unsure.

My recommendation would be to create/maintain an specific container image with all the prerequisites. Then, use that image both in CI and locally. That will reduce startup time, and it will provide exactly the same result for all developers/collaborators and CI services.

It might make sense to create litex-hub/containers or symbiflow/containers for maintaining build images. We own the symbiflow namespace in dockerhub (I can transfer/give access to any of you). See ghdl/docker (and ghdl/ghdl) and hdl/containers for practical examples.

ajelinski replied:
All of the commands provided have to be executed in the same OS and are the same regardless whether used in a host OS or a container. The commands have been tested mostly on Debian/Ubuntu, but apart from the package installation tool there shouldn't be many problems to use it on many other GNU/Linux OSs.

Did command separation into two blocks made you think that some commands should be executed in a host OS and the others in a container? In the same sentence where you suggested a change there is: 
> The commands from the following subsections are enough to build a package in a clean docker container based on the *ubuntu* Docker image.

but maybe I should stress that the commands should also be executed there? Maybe you want to suggest some change to make everything clear?

Your recommendation seems to be very reasonable but this would be quite a change. I don't think it should be handled by this PR. Maybe you could create an issue pointing out what are the problems with the current situation and your suggestion? Simply to not forget about it as I think it is a good idea.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+The tool is also used within the CI, which makes the locally built
+package much more similar to the one built by the CI workflow.
+
+The tool is a Python module.
```
I would suggest being more explicit than "*the tool*". Maybe "*the entrypoint to this build tooling is a Python module with a CLI*".

ajelinski replied:
You're right but I changed it to "The `conda-build-prepare` is a Python module with a CLI."

ajelinski replied:
You might've been confused about which tool this was as there was "the tool". I made the whole paragraph much more precise now.

umarcor commented on README.md:
```diff
@@ -89,61 +89,143 @@ builds like `gcc` generate a *lot* of output), and builds timing out
+package much more similar to the one built by the CI workflow.
+
+The tool is a Python module.
+The only required arguments are the path to the recipe corresponding to the
```
Maybe out of scope of this PR, but it'd be interesting to generate this details as a canonical help (docstrings) in Python. Then, copy the output of `-h` here.

Paebbels/pyAttributes might be useful.

ajelinski replied:
I don't think that users should really be informed about all the details of `conda-build-prepare`. If the primary goal is to make building process similar to the CI one then it's best when the same arguments as in that CI are used.

Didn't know the pyAttributes though, I'll check it out.

ajelinski replied:
You might've been confused about what tool this was as there was "the tool". I made the whole paragraph much more precise now.

umarcor replied:
My point is, are users expected to execute `conda-build-prepare` or is it for developers only? If it's going to be executed by users, I think that a list output is desirable:

- `--dir`: ...
- `--channels`: ...
- `--packages`: ...

By using pyAttributes or a similar solution, the amount of info shown to users can be decided.

ajelinski replied:
Anybody who wants to build a package as it's done in the CI should use *conda-build-prepare* and with the `--channels` and `--packages` the same as in README (=CI).

I've changed the whole section on *conda-build-prepare* arguments and I think it's much more comprehensible now. I've explained thoroughly the options that user might be interested in, what should be passed to the remaining ones and why.

ajelinski commented:
Updated `conda-build-prepare` version to the one that automatically provides `DATE_NUM` and `DATE_STR` variables.

I've added the CI `tuttest` workflow that tests commands from `README.md`. The first job gathers the commands and six subsequent jobs test them in the Docker containers with three latest Debian and Ubuntu versions.

It won't be run if none of the `README.md`, `tuttest.yml` and `binutils/` changes. Also, now the `build-packages` CI won't be run if only `README.md` or `tuttest.yml` changes.

umarcor commented on .github/workflows/tuttest.yml:
```diff
@@ -0,0 +1,102 @@
+
+      - name: Add checkout commands for a Pull Request CI
+        if: ${{ github.event_name == 'pull_request' }}
+        run: |
```
Maybe this step and the next one can be merged?

```sh
_branch="$GITHUB_SHA"
if [ "${{ github.event_name }}" = "pull_request" ]; then
  _branch="ci-pr-branch"
  echo "git fetch origin $GITHUB_REF:${_branch}" >>$SCRIPT
fi
echo "git checkout ${_branch}" >>$SCRIPT
```

umarcor commented on .github/workflows/tuttest.yml:
```diff
@@ -0,0 +1,102 @@
+jobs:
+  grab-readme-commands:
+    runs-on: ubuntu-latest
+    defaults:
```
I think that setting the default shell is not required here.

umarcor commented on .github/workflows/tuttest.yml:
```diff
@@ -0,0 +1,102 @@
+
+      - uses: actions/upload-artifact@v2
+        with:
+          name: ${{ env.ARTIFACT_NAME }}
```
Since there is a single artifact, setting the name is not necessary.

ajelinski replied:
It's good to know it isn't needed but I've left it unchanged because without a name all artifacts are downloaded to separate directories, even if there's a single one. I'd need to adapt to it or specify path during downloading so I think the current solution is better.

umarcor commented on .github/workflows/tuttest.yml:
```diff
@@ -0,0 +1,102 @@
+    steps:
+      - uses: actions/download-artifact@v2
+        with:
+          name: ${{ env.ARTIFACT_NAME }}
```
You can optionally set the `path` argument.

ajelinski commented:
I think it's now polished to the limit and could be merged. Any further remarks @mithro @PiotrZierhoffer @umarcor?

mithro reviewed (approved):
This all looks really good!

ajelinski commented:
Fixed two minor problems (one of them was breaking the CI on a fork), made some reasonable commits out of the WIP ones and rebased on top of the current master.

ajelinski commented:
`tuttest` workflow is green and the `linux-musl` has been already tested before so I'm merging this PR.
